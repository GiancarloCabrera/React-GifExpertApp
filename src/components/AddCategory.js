import React, { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategories }) => {
                                //Se pone un string vacio, porque si no se pone nada
                                // queda como undefined y falla
    const [inputValue, setInputValue] = useState('');
    const handleInput = (e) => {
        setInputValue(e.target.value)
        console.log(e.target.value);
    }
    const handleSubmit = (e) => {
        //Esto se usa para prevenir que el fromulario se refresque solo
        e.preventDefault();
        console.log('Formulario Hecho!!!!')
        //trim borra todos los espacios de un string
        //Si inputValue tiene mas de dos valores
        if (inputValue.trim().length > 2) {
            //El setCategories trae por defecto los valores que fueron incluidos cuando
            //el setCategories fue creado
            setCategories(cats => [inputValue, ...cats]);//inputValue contiene el valor insertado en el input
            //Ese cats contiene todos los valores almacenados hasta el momento
            //setInputValue('');
        }
    }
  return (
    //El formulario se sube cuando se presione enter
    <form onSubmit={ handleSubmit }>
        <h1>{ inputValue }</h1>
        {/* Se llama el input, el cual su valor sera lo que contenga
            el inputValue, onChange, osea cada vez que haya un cambio en ese input, se llama la funcion
            la cual le va a setear el nuevo valor escrito al inputValue, encargado de almacenar la info
            del input */}
        <input
            type="text"
            value={ inputValue }
            onChange={ handleInput }
        />
    </form>
  )
}

//La funcion setCategories es requerida obligatoriamente

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}