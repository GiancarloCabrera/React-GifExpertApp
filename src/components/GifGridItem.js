import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({ img }) => {
  console.log('IMG:' ,img);
  return (
    <div className='card animate__animated animate__bounce'>
        <img src= { img.url } alt= { img.title }/>
        <p> { img.title }</p>
    </div>
  )
}

GifGridItem.propTypes = {
  img: PropTypes.object.isRequired
}
//ENZYME REQUIERE SIEMRPE QUE LOS COMPONENTES A EVALUAR SEAN EXPORTADOS 
// POR DEFAULT
export default GifGridItem;