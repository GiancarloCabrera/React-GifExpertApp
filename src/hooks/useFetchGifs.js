// HOOKS
//Se pone useFetch en el nombre del archivo para saber que es un hook

import { useEffect, useState } from "react"
import { getGifs } from "../helpers/GetGifs"
                        //IMPORTANTEEEEEEEEEEEE
                        //Como no es un componente sino una funcion, no usamos desestructuracion a menos que los necesitemos
export const useFetchGifs = ( category ) => {
    const [state, setState] = useState( {
        data: [],
        loading: true
    })

    //useEffect
    //Voy a implementar useEffecto debido a que quiero que la la request http
    // solo corra una sola vez
    //No hay argumentos en el useEffect asi se vayan a usar adentro
    useEffect( () => { //Los efectos no pueden ser async
        getGifs( category ).then( imgs => {
            //imgs trae toda la data de la api requerida por el usuario
            //Llamamos setState para cambiar el estado del objeto state
            setState({
                //Una vez cargado cambiamos los estados del objeto
                data: imgs, //La data sera las imagenes que trae nuestra request http
                loading: false //Como ya llegan las fotos, debe dejar de cargar
            });
        })
    },[ category ])//Arreglo de dependencias

    return state;
}