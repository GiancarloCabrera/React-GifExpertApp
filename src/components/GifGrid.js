import React from 'react'
//import { GifGridItem } from './GifGridItem';
import { PropTypes } from 'prop-types';
//import { getGifs } from '../helpers/GetGifs'; //Se importa de helpers la funcion
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //TODO ESTE CODIGO COMENTADO FUE MOVIDO A useFetchGifs DEBIDO A QUE ES UNA BUENA PRACTICA
    //ALMACENARLO EN UN HOOK

    //Creo un estado el cual sera imagenes, este recibira la info del
    // objeto gifs la cual sera mostrada por pantalla
    //const [images, setImages] = useState([]);

    //useEffect
    //Muchas veces hay codigo que necesito que corra una sola vez, asi haya cambios en el
    // en el componente. Para eso se usa el useEffect
    //En este caso react esta haciendo la request a la api cada vez que hay un camnio en el componente, y
    // yo no quiero que sea asi, necesito que solo haga la request una vez y ya, por ende voy a meter la funcion getGifs dentro
    // del hook useEffect

    // useEffect( () => {
    //     getGifs( category ).then( imgs => { //Como la funcion retorna un objeto, yo puedo acceder a el y mandarlo al setImages
    //         setImages (imgs);
    //     });
    // }, [ category ]); //Dentro de ese array van las dependencias que quiero incluir, pongo category porque cada vez que cambie quiero que ese codigo vuelva y corra

    //Esta funcion es importada del hook del archivo useFectchGifs y lo desestructuro
    const { data, loading } = useFetchGifs( category );
                        //Asi se manda una variable de este componente a otro

    return (
        <>
            <h3 className='animate__animated animate_fadeIn'>{ category }</h3>

            {/* Llamamos loading y hacemos un operador ternario,
            si loading es true muestre el parrafo, si no, no haga nada*/}
            
            { loading && <p className='animate__animated animate_flash'>LOADING</p> }

            <div className='card-grid'>
                {   //Llamamos el array data que tiene la info que queremos de la api y lo mapeamos para que mueste todo lo que queremos
                    data.map( img => {
                        //Tenemos un nuevo componente el cual esta encargado de mostrar las imagenes con sus titulos
                        return <GifGridItem 
                            key={ img.id }
                            img = { img }/>
                    })
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
