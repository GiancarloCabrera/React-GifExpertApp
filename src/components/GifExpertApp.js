import React, { useState } from "react";
import { AddCategory } from "./AddCategory";
import { GifGrid } from "./GifGrid";

const GitExpertApp = () => {

    //const categories = ['One Punch', 'Samurai X', 'Dragon Ball'];
    //Como el array puede cambiar, puede tener mas o menos, usamos el useState
    const [categories, setCategories]= useState(['Samurai X']);
    
    //const handleAdd = () => {

        //PARA AGREGAR ELEMENTOS AL ARRAY USANDO setCategories
        //Se usa el operador spread. mantiene lo anterior y anade lo nuevo
        //setCategories([...categories, 'FIGHTER']);
        console.log(categories)

        //OTRA MANERA USANDO --------- CALLBACK
        //El cats contiene lo ya guardado y se le adiciona el LALALAND
        //setCategories(cats => [...cats, 'LALALAND'])

        //SPREADDDDDDDDDDDDDD
        //Concierte un array a una lista
    //}
    console.log(categories)

    return (
        <div>
            <h2>GitExpertApp</h2>
            {/* COMUNICACION ENTRE COMPONENTESSSSSSSSSSSSSSSSS
                Por medio del props se puede llamar el setCategories el cual 
                va a actualizar el estado del array */}

            <AddCategory setCategories= { setCategories }/>
            {/* Ese hr es una linea */}
            <hr></hr>
    
            

            {/* Mostrar un array de forma ordenada tipo lista,
                se pone la propiedad ol que es usada para listas */}
            <ol>
                {/* Con el map. lo que hace es recorrer el array y retornar
                    cada uno de los elementos del array */}
                {   //El map recibe 2 argumentos,
                    //el elemento del array y su indice
                    //No siempre es necesario poner el indice
                    categories.map( category => { 
                        //Antes se llamaba la categoria del array para mostrarla,
                        //ahora se creo un nuevo componente el cual va a mostar la categoria
                        //del array y hara una peticion http para que muestre gifs de la categoria
                        //return <li key={ category }> { category } </li>

                        //Este es el nuevo componente
                        //Recordar que REACT siempre pide un key para los valores 
                        // para poder diferenciarlos
                        return  <GifGrid 
                            key={ category }
                            category={ category }
                        />
                    })
                }
            </ol>
        </div>
    );
}

export default GitExpertApp;