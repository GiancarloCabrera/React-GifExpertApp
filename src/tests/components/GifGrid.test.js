import React from "react";
import { shallow } from "enzyme"
import { GifGrid } from "../../components/GifGrid"
import { useFetchGifs } from "../../hooks/useFetchGifs";
import '@testing-library/jest-dom';

//Con este jest.mock fingimos el funcionamiento de esa funcion que es un hook
jest.mock('../../hooks/useFetchGifs')

describe('Pruebas componente GifGrid', () => { 

    const category = 'Shadow Hunters';

    test('El componente debe ser mostrado de manera correcta', () => { 
        //Hacemos un llamdo de la funcion y hacemos que retorne este objeto
        //Cabe recordar que la funcion requiere devolver algo, por eso se 
        // esta llamando el mockReturnValue
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const app = shallow(<GifGrid category={ category }/>);
        expect( app ).toMatchSnapshot();
    });    

    test('Debe de mostrar items cuando se cargan imagenes, useFetchGifs', () => { 
        
        //Con esto fingimos que se haya llamado la funcion y haya retornado esto
        const gifs = [{
            id: 'ABC',
            url: 'https://noon.com',
            title: 'HEYYYYYYY'
        },
        {
            id: 'CDJ',
            url: 'https://lalalal.com',
            title: 'OEEEEEEEE'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const app = shallow(<GifGrid category={ category }/>);

        //Evaluar que el componente se esta mostrando bien
        expect( app ).toMatchSnapshot();

        //Solo hay un parrafo y solo aparece si loading esta en true, por ende
        // no debe aparecer porque loading es false por defecto
        //Asi se evalua
        expect( app.find('p').exists() ).toBe(false);

        //Evaluar si el componente esta recibiendo el mismo numero de 
        // objetos que la funcion useFetch esta retornando
        //Se busca el componente en el html y se espera que tenga el mismo tmaano que el array que yo le estoy pasando
        expect( app.find('GifGridItem').length).toBe( gifs.length );
    });
})