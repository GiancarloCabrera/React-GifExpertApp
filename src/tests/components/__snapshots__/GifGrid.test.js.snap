// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas componente GifGrid Debe de mostrar items cuando se cargan imagenes, useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate_fadeIn"
  >
    Shadow Hunters
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      img={
        Object {
          "id": "ABC",
          "title": "HEYYYYYYY",
          "url": "https://noon.com",
        }
      }
      key="ABC"
    />
    <GifGridItem
      img={
        Object {
          "id": "CDJ",
          "title": "OEEEEEEEE",
          "url": "https://lalalal.com",
        }
      }
      key="CDJ"
    />
  </div>
</Fragment>
`;

exports[`Pruebas componente GifGrid El componente debe ser mostrado de manera correcta 1`] = `
<Fragment>
  <h3
    className="animate__animated animate_fadeIn"
  >
    Shadow Hunters
  </h3>
  <p
    className="animate__animated animate_flash"
  >
    LOADING
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
